lib/RTCSession.js

```diff
@@ -626,45 +626,53 @@ module.exports = class RTCSession extends EventEmitter
     // TODO: This may throw an error, should react.
     this._createRTCConnection(pcConfig, rtcConstraints);
 
-    // If a local MediaStream is given use it.
-    if (mediaStream)
-    {
-      userMediaSucceeded.call(this, mediaStream);
-    }
-    // If at least audio or video is requested prompt getUserMedia.
-    else if (mediaConstraints.audio || mediaConstraints.video)
-    {
-      this._localMediaStreamLocallyGenerated = true;
-      navigator.mediaDevices.getUserMedia(mediaConstraints)
-        .then(userMediaSucceeded.bind(this))
-        .catch((error) =>
+    Promise.resolve()
+      // Handle local MediaStream.
+      .then(() =>
+      {
+        // A local MediaStream is given, use it.
+        if (mediaStream)
         {
-          userMediaFailed.call(this, error);
-
-          debugerror('emit "getusermediafailed" [error:%o]', error);
+          return mediaStream;
+        }
 
-          this.emit('getusermediafailed', error);
-        });
-    // Otherwise don't prompt getUserMedia.
-    }
-    else
-    {
-      userMediaSucceeded.call(this, null);
-    }
+        // Audio and/or video requested, prompt getUserMedia.
+        else if (mediaConstraints.audio || mediaConstraints.video)
+        {
+          this._localMediaStreamLocallyGenerated = true;
 
-    // User media succeeded.
-    function userMediaSucceeded(stream)
-    {
-      if (this._status === C.STATUS_TERMINATED) { return; }
+          return navigator.mediaDevices.getUserMedia(mediaConstraints)
+            .catch(() =>
+            {
+              request.reply(480);
+              this._failed('local', null, JsSIP_C.causes.USER_DENIED_MEDIA_ACCESS);
 
-      this._localMediaStream = stream;
-      if (stream)
+              return Promise.reject('getUserMedia');
```

Comment:
IMHO it's better to do `throw new Error('getUserMedia() failed');` here rather than returning a rejected promise.

Reply:
Note: I reject the Promise with a `string` and not an `Error` because it is only used by the `catch()` statement on the Promise chain in order to make sure that the error has been handled by the corresponding block. It is just a label (I called it **step**) so the `catch` statement can compare it against the well known steps that may have failed (and already handled the error).

See [here](https://github.com/versatica/JsSIP/blob/promisify/lib/RTCSession.js#L760)

Having said that, the following could be more clean and generic:

```javascript
throw new Exceptions.PromiseChainError('getUserMedia');
```

```javascript
.catch((error) =>
{
  if (! (error instanceof Excepcions.PromiseChainError))
 {
   throw new Error(`Unhandled exception: ${error}`);
 }
});
```